config/src/config/node_config_loader.rs

```diff
@@ -16,6 +16,45 @@ use aptos_types::{
 };
 use std::path::Path;
 
+/// A simple enum to represent the type of a node
+/// as determined from the config file.
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum NodeType {
+    Validator,
+    ValidatorFullnode,
+    PublicFullnode,
+}
+
+impl NodeType {
+    pub fn is_validator(self) -> bool {
+        self == NodeType::Validator
+    }
+
+    pub fn is_validator_fullnode(self) -> bool {
+        self == NodeType::ValidatorFullnode
+    }
+
+    /// Returns the type of the node as determined by the node config
+    pub fn extract_from_config(node_config: &NodeConfig) -> Self {
+        // Validator nodes are trivial to detect
+        if node_config.base.role.is_validator() {
+            return NodeType::Validator;
+        }
+
+        // Otherwise, we must decipher between VFNs and PFNs
+        // based on the presence of a VFN network.
+        let vfn_network_found = node_config
+            .full_node_networks
+            .iter()
+            .any(|network| network.network_id.is_vfn_network());
+        if vfn_network_found {
+            NodeType::ValidatorFullnode
+        } else {
+            NodeType::PublicFullnode
+        }
+    }
```

Comment:
This is kinda strange there's now a Config type and an internal type

Reply:
Yeah, the internal type is technically what we want (e.g., it makes the node configs super clear), but to maintain backward compatibility, I couldn't change the existing config type (i.e., `RoleType`). So, I've just defined the internal locally (and it's never exposed).

If we want to migrate to the external type eventually, we'll need to take a pass over all of our docs and our various template files to add a new `validator_fullnode` entry. But, I don't want to do that here (or now) 😄